### ISSUE TITLE
Regression in dill >= 0.3.5: cannot pickle 'KeyedRef' object

### ISSUE THREAD

**Issue by schlegelp** [uqfoundation/dill#497]
Hi! I've been running into some strange behaviour using pathos (0.2.8 or 0.2.9) with dill 0.3.5 and 0.3.5.1 (observed on both OSX and Ubuntu)

I'm working with fairly complex data objects and I couldn't nail down where exactly things go wrong. I'm afraid I can't produce a useful minimal example but perhaps the error message helps:

```Python
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [6], in <cell line: 1>()
----> 1 navis.prune_twigs(pr, '1 micron', parallel=True)

File ~/dev/Python/navis_error/navis/navis/utils/misc.py:238, in map_neuronlist.<locals>.decorator.<locals>.wrapper(*args, **kwargs)
    228 proc = core.NeuronProcessor(nl, function,
    229                             parallel=parallel,
    230                             desc=desc,
   (...)
    235                             exclude_zip=excl,
    236                             n_cores=n_cores)
    237 # Apply function
--> 238 res = proc(nl, *args, **kwargs)
    240 # When using parallel processing, the neurons will not actually
    241 # have been modified inplace - in that case we will simply
    242 # replace the neurons in `nl`
    243 if inplace:

File ~/dev/Python/navis_error/navis/navis/core/core_utils.py:407, in NeuronProcessor.__call__(self, *args, **kwargs)
    399             wrapper = _try_call
    401         #import pickle 
    402         #_ = pickle.dumps(wrapper) 
    403 
    404         #for c in combinations:
    405         #   _ = pickle.dumps(c)
--> 407         res = list(config.tqdm(pool.imap(wrapper,
    408                                          combinations,
    409                                          chunksize=chunksize),
    410                                total=len(combinations),
    411                                desc=self.desc,
    412                                disable=config.pbar_hide or not self.progress,
    413                                leave=config.pbar_leave))
    414 else:
    415     res = []

File ~/.pyenv/versions/3.9.5/lib/python3.9/site-packages/tqdm/std.py:1195, in tqdm.__iter__(self)
   1192 time = self._time
   1194 try:
-> 1195     for obj in iterable:
   1196         yield obj
   1197         # Update and possibly print the progressbar.
   1198         # Note: does not call self.update(1) for speed optimisation.

File ~/.pyenv/versions/3.9.5/lib/python3.9/site-packages/multiprocess/pool.py:870, in IMapIterator.next(self, timeout)
    868 if success:
    869     return value
--> 870 raise value

File ~/.pyenv/versions/3.9.5/lib/python3.9/site-packages/multiprocess/pool.py:537, in Pool._handle_tasks(taskqueue, put, outqueue, pool, cache)
    535     break
    536 try:
--> 537     put(task)
    538 except Exception as e:
    539     job, idx = task[:2]

File ~/.pyenv/versions/3.9.5/lib/python3.9/site-packages/multiprocess/connection.py:214, in _ConnectionBase.send(self, obj)
    212 self._check_closed()
    213 self._check_writable()
--> 214 self._send_bytes(_ForkingPickler.dumps(obj))

File ~/.pyenv/versions/3.9.5/lib/python3.9/site-packages/multiprocess/reduction.py:54, in ForkingPickler.dumps(cls, obj, protocol, *args, **kwds)
     51 @classmethod
     52 def dumps(cls, obj, protocol=None, *args, **kwds):
     53     buf = io.BytesIO()
---> 54     cls(buf, protocol, *args, **kwds).dump(obj)
     55     return buf.getbuffer()

File ~/.pyenv/versions/3.9.5/lib/python3.9/site-packages/dill/_dill.py:620, in Pickler.dump(self, obj)
    618     raise PicklingError(msg)
    619 else:
--> 620     StockPickler.dump(self, obj)
    621 return

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:487, in _Pickler.dump(self, obj)
    485 if self.proto >= 4:
    486     self.framer.start_framing()
--> 487 self.save(obj)
    488 self.write(STOP)
    489 self.framer.end_framing()

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:901, in _Pickler.save_tuple(self, obj)
    899 write(MARK)
    900 for element in obj:
--> 901     save(element)
    903 if id(obj) in memo:
    904     # Subtle.  d was not in memo when we entered save_tuple(), so
    905     # the process of saving the tuple's elements must have saved
   (...)
    909     # could have been done in the "for element" loop instead, but
    910     # recursive tuples are a rare thing.
    911     get = self.get(memo[id(obj)][0])

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:886, in _Pickler.save_tuple(self, obj)
    884 if n <= 3 and self.proto >= 2:
    885     for element in obj:
--> 886         save(element)
    887     # Subtle.  Same as in the big comment below.
    888     if id(obj) in memo:

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:886, in _Pickler.save_tuple(self, obj)
    884 if n <= 3 and self.proto >= 2:
    885     for element in obj:
--> 886         save(element)
    887     # Subtle.  Same as in the big comment below.
    888     if id(obj) in memo:

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:886, in _Pickler.save_tuple(self, obj)
    884 if n <= 3 and self.proto >= 2:
    885     for element in obj:
--> 886         save(element)
    887     # Subtle.  Same as in the big comment below.
    888     if id(obj) in memo:

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File ~/.pyenv/versions/3.9.5/lib/python3.9/site-packages/dill/_dill.py:1963, in save_function(pickler, obj)
   1960     if state_dict:
   1961         state = state, state_dict
-> 1963     _save_with_postproc(pickler, (_create_function, (
   1964           obj.__code__, globs, obj.__name__, obj.__defaults__,
   1965           closure
   1966     ), state), obj=obj, postproc_list=postproc_list)
   1967 else:
   1968     closure = obj.func_closure

File ~/.pyenv/versions/3.9.5/lib/python3.9/site-packages/dill/_dill.py:1154, in _save_with_postproc(pickler, reduction, is_pickler_dill, obj, postproc_list)
   1152 if source:
   1153     pickler.write(pickler.get(pickler.memo[id(dest)][0]))
-> 1154     pickler._batch_setitems(iter(source.items()))
   1155 else:
   1156     # Updating with an empty dictionary. Same as doing nothing.
   1157     continue

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:997, in _Pickler._batch_setitems(self, items)
    995     for k, v in tmp:
    996         save(k)
--> 997         save(v)
    998     write(SETITEMS)
    999 elif n:

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File ~/.pyenv/versions/3.9.5/lib/python3.9/site-packages/dill/_dill.py:1765, in save_module(pickler, obj)
   1762     _main_dict = obj.__dict__.copy() #XXX: better no copy? option to copy?
   1763     [_main_dict.pop(item, None) for item in singletontypes
   1764         + ["__builtins__", "__loader__"]]
-> 1765     pickler.save_reduce(_import_module, (obj.__name__,), obj=obj,
   1766                         state=_main_dict)
   1767     log.info("# M1")
   1768 elif PY3 and obj.__name__ == "dill._dill":

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:717, in _Pickler.save_reduce(self, func, args, state, listitems, dictitems, state_setter, obj)
    715 if state is not None:
    716     if state_setter is None:
--> 717         save(state)
    718         write(BUILD)
    719     else:
    720         # If a state_setter is specified, call it instead of load_build
    721         # to update obj's with its previous state.
    722         # First, push state_setter and its tuple of expected arguments
    723         # (obj, state) onto the stack.

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File ~/.pyenv/versions/3.9.5/lib/python3.9/site-packages/dill/_dill.py:1251, in save_module_dict(pickler, obj)
   1248     if is_dill(pickler, child=False) and pickler._session:
   1249         # we only care about session the first pass thru
   1250         pickler._first_pass = False
-> 1251     StockPickler.save_dict(pickler, obj)
   1252     log.info("# D2")
   1253 return

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:971, in _Pickler.save_dict(self, obj)
    968     self.write(MARK + DICT)
    970 self.memoize(obj)
--> 971 self._batch_setitems(obj.items())

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:997, in _Pickler._batch_setitems(self, items)
    995     for k, v in tmp:
    996         save(k)
--> 997         save(v)
    998     write(SETITEMS)
    999 elif n:

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:603, in _Pickler.save(self, obj, save_persistent_id)
    599     raise PicklingError("Tuple returned by %s must have "
    600                         "two to six elements" % reduce)
    602 # Save the reduce() output and finally memoize the object
--> 603 self.save_reduce(obj=obj, *rv)

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:717, in _Pickler.save_reduce(self, func, args, state, listitems, dictitems, state_setter, obj)
    715 if state is not None:
    716     if state_setter is None:
--> 717         save(state)
    718         write(BUILD)
    719     else:
    720         # If a state_setter is specified, call it instead of load_build
    721         # to update obj's with its previous state.
    722         # First, push state_setter and its tuple of expected arguments
    723         # (obj, state) onto the stack.

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File ~/.pyenv/versions/3.9.5/lib/python3.9/site-packages/dill/_dill.py:1251, in save_module_dict(pickler, obj)
   1248     if is_dill(pickler, child=False) and pickler._session:
   1249         # we only care about session the first pass thru
   1250         pickler._first_pass = False
-> 1251     StockPickler.save_dict(pickler, obj)
   1252     log.info("# D2")
   1253 return

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:971, in _Pickler.save_dict(self, obj)
    968     self.write(MARK + DICT)
    970 self.memoize(obj)
--> 971 self._batch_setitems(obj.items())

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:997, in _Pickler._batch_setitems(self, items)
    995     for k, v in tmp:
    996         save(k)
--> 997         save(v)
    998     write(SETITEMS)
    999 elif n:

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File ~/.pyenv/versions/3.9.5/lib/python3.9/site-packages/dill/_dill.py:1251, in save_module_dict(pickler, obj)
   1248     if is_dill(pickler, child=False) and pickler._session:
   1249         # we only care about session the first pass thru
   1250         pickler._first_pass = False
-> 1251     StockPickler.save_dict(pickler, obj)
   1252     log.info("# D2")
   1253 return

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:971, in _Pickler.save_dict(self, obj)
    968     self.write(MARK + DICT)
    970 self.memoize(obj)
--> 971 self._batch_setitems(obj.items())

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:997, in _Pickler._batch_setitems(self, items)
    995     for k, v in tmp:
    996         save(k)
--> 997         save(v)
    998     write(SETITEMS)
    999 elif n:

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:603, in _Pickler.save(self, obj, save_persistent_id)
    599     raise PicklingError("Tuple returned by %s must have "
    600                         "two to six elements" % reduce)
    602 # Save the reduce() output and finally memoize the object
--> 603 self.save_reduce(obj=obj, *rv)

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:717, in _Pickler.save_reduce(self, func, args, state, listitems, dictitems, state_setter, obj)
    715 if state is not None:
    716     if state_setter is None:
--> 717         save(state)
    718         write(BUILD)
    719     else:
    720         # If a state_setter is specified, call it instead of load_build
    721         # to update obj's with its previous state.
    722         # First, push state_setter and its tuple of expected arguments
    723         # (obj, state) onto the stack.

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File ~/.pyenv/versions/3.9.5/lib/python3.9/site-packages/dill/_dill.py:1251, in save_module_dict(pickler, obj)
   1248     if is_dill(pickler, child=False) and pickler._session:
   1249         # we only care about session the first pass thru
   1250         pickler._first_pass = False
-> 1251     StockPickler.save_dict(pickler, obj)
   1252     log.info("# D2")
   1253 return

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:971, in _Pickler.save_dict(self, obj)
    968     self.write(MARK + DICT)
    970 self.memoize(obj)
--> 971 self._batch_setitems(obj.items())

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:997, in _Pickler._batch_setitems(self, items)
    995     for k, v in tmp:
    996         save(k)
--> 997         save(v)
    998     write(SETITEMS)
    999 elif n:

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:603, in _Pickler.save(self, obj, save_persistent_id)
    599     raise PicklingError("Tuple returned by %s must have "
    600                         "two to six elements" % reduce)
    602 # Save the reduce() output and finally memoize the object
--> 603 self.save_reduce(obj=obj, *rv)

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:717, in _Pickler.save_reduce(self, func, args, state, listitems, dictitems, state_setter, obj)
    715 if state is not None:
    716     if state_setter is None:
--> 717         save(state)
    718         write(BUILD)
    719     else:
    720         # If a state_setter is specified, call it instead of load_build
    721         # to update obj's with its previous state.
    722         # First, push state_setter and its tuple of expected arguments
    723         # (obj, state) onto the stack.

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File ~/.pyenv/versions/3.9.5/lib/python3.9/site-packages/dill/_dill.py:1251, in save_module_dict(pickler, obj)
   1248     if is_dill(pickler, child=False) and pickler._session:
   1249         # we only care about session the first pass thru
   1250         pickler._first_pass = False
-> 1251     StockPickler.save_dict(pickler, obj)
   1252     log.info("# D2")
   1253 return

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:971, in _Pickler.save_dict(self, obj)
    968     self.write(MARK + DICT)
    970 self.memoize(obj)
--> 971 self._batch_setitems(obj.items())

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:997, in _Pickler._batch_setitems(self, items)
    995     for k, v in tmp:
    996         save(k)
--> 997         save(v)
    998     write(SETITEMS)
    999 elif n:

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File ~/.pyenv/versions/3.9.5/lib/python3.9/site-packages/dill/_dill.py:1251, in save_module_dict(pickler, obj)
   1248     if is_dill(pickler, child=False) and pickler._session:
   1249         # we only care about session the first pass thru
   1250         pickler._first_pass = False
-> 1251     StockPickler.save_dict(pickler, obj)
   1252     log.info("# D2")
   1253 return

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:971, in _Pickler.save_dict(self, obj)
    968     self.write(MARK + DICT)
    970 self.memoize(obj)
--> 971 self._batch_setitems(obj.items())

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:997, in _Pickler._batch_setitems(self, items)
    995     for k, v in tmp:
    996         save(k)
--> 997         save(v)
    998     write(SETITEMS)
    999 elif n:

File ~/.pyenv/versions/3.9.5/lib/python3.9/pickle.py:578, in _Pickler.save(self, obj, save_persistent_id)
    576 reduce = getattr(obj, "__reduce_ex__", None)
    577 if reduce is not None:
--> 578     rv = reduce(self.proto)
    579 else:
    580     reduce = getattr(obj, "__reduce__", None)

TypeError: cannot pickle 'KeyedRef' object
```

I also tried logging via `dill.detect.trace(True)` as suggested in an earlier thread:

```
F1: <function starargs.<locals>.<lambda> at 0x7f48012d4ca0>
F2: <function _create_function at 0x7f481ebdaee0>
# F2
Co: <code object <lambda> at 0x7f480209cbe0, file "/home/jefferis/.pyenv/versions/3.9.5/lib/python3.9/site-packages/pathos/helpers/mp_helper.py", line 15>
F2: <function _create_code at 0x7f481ebdaf70>
# F2
# Co
D4: <dict object at 0x7f48020d1f80>
# D4
Ce2: <cell at 0x7f47f891ea60: function object at 0x7f4802058b80>
F2: <function _create_cell at 0x7f481ebdd3a0>
# F2
# Ce2
D2: <dict object at 0x7f47f8c3ab80>
# D2
D2: <dict object at 0x7f47ef5d1e00>
D2: <dict object at 0x7f47f8946ec0>
# D2
# D2
M2: <module 'dill._dill' from '/home/jefferis/.pyenv/versions/3.9.5/lib/python3.9/site-packages/dill/_dill.py'>
# M2
F2: <function _call at 0x7f4802058b80>
# F2
# F1
F1: <function prune_twigs at 0x7f4802124940>
Co: <code object wrapper at 0x7f4809b64030, file "/home/jefferis/dev/Python/navis_error/navis/navis/utils/misc.py", line 473>
Co: <code object <listcomp> at 0x7f4809b5ff50, file "/home/jefferis/dev/Python/navis_error/navis/navis/utils/misc.py", line 567>
# Co
# Co
D2: <dict object at 0x7f47ef5d17c0>
# D2
Ce2: <cell at 0x7f48021238b0: list object at 0x7f48036cf2c0>
# Ce2
Ce2: <cell at 0x7f48021238e0: dict object at 0x7f48095b9280>
# Ce2
Ce2: <cell at 0x7f4802123a00: function object at 0x7f48021248b0>
# Ce2
Ce2: <cell at 0x7f4802123910: bool object at 0x562294a25060>
# Ce2
Ce2: <cell at 0x7f4802123940: bool object at 0x562294a25060>
# Ce2
Ce2: <cell at 0x7f4802123970: str object at 0x7f481f81ec70>
# Ce2
Ce2: <cell at 0x7f48021239a0: list object at 0x7f48036c8480>
# Ce2
Ce2: <cell at 0x7f48021239d0: bool object at 0x562294a25060>
# Ce2
D2: <dict object at 0x7f4802128040>
F1: <function prune_twigs at 0x7f48021248b0>
Co: <code object prune_twigs at 0x7f48021190e0, file "/home/jefferis/dev/Python/navis_error/navis/navis/morpho/manipulation.py", line 265>
# Co
D4: <dict object at 0x7f4802418280>
# D4
D2: <dict object at 0x7f48021280c0>
# D2
D2: <dict object at 0x7f47f899e440>
D2: <dict object at 0x7f4802122140>
T4: <class 'typing.ForwardRef'>
# T4
D2: <dict object at 0x7f47f8996280>
Co: <code object <module> at 0x7f480211ff50, file "<string>", line 1>
# Co
# D2
D2: <dict object at 0x7f47f8bcf6c0>
Co: <code object <module> at 0x7f48021263a0, file "<string>", line 1>
# Co
# D2
T1: <class 'float'>
F2: <function _load_type at 0x7f481ebdadc0>
# F2
# T1
T1: <class 'str'>
# T1
T1: <class 'bool'>
# T1
T1: <class 'int'>
# T1
# D2
# D2
# F1
# D2
D2: <dict object at 0x7f47ef5e8500>
# D2
D2: <dict object at 0x7f48095b9280>
# D2
T4: <class '_frozen_importlib_external.SourceFileLoader'>
# T4
D2: <dict object at 0x7f480ce7bec0>
# D2
T4: <class '_frozen_importlib.ModuleSpec'>
# T4
D2: <dict object at 0x7f480ce7f600>
# D2
D4: <dict object at 0x7f4821dfadc0>
# D4
M2: <module 'inspect' from '/home/jefferis/.pyenv/versions/3.9.5/lib/python3.9/inspect.py'>
F2: <function _import_module at 0x7f481ebddc10>
# F2
# M2
M2: <module 'math' from '/home/jefferis/.pyenv/versions/3.9.5/lib/python3.9/lib-dynload/math.cpython-39-x86_64-linux-gnu.so'>
# M2
M2: <module 'os' from '/home/jefferis/.pyenv/versions/3.9.5/lib/python3.9/os.py'>
# M2
M2: <module 're' from '/home/jefferis/.pyenv/versions/3.9.5/lib/python3.9/re.py'>
# M2
M2: <module 'requests' from '/home/jefferis/.pyenv/versions/3.9.5/lib/python3.9/site-packages/requests/__init__.py'>
# M2
M2: <module 'sys' (built-in)>
# M2
M2: <module 'urllib' from '/home/jefferis/.pyenv/versions/3.9.5/lib/python3.9/urllib/__init__.py'>
# M2
M2: <module 'numpy' from '/home/jefferis/.pyenv/versions/3.9.5/lib/python3.9/site-packages/numpy/__init__.py'>
# M2
M2: <module 'pandas' from '/home/jefferis/.pyenv/versions/3.9.5/lib/python3.9/site-packages/pandas/__init__.py'>
# M2
F2: <function wraps at 0x7f4821c74040>
# F2
F2: <function dedent at 0x7f48217ef160>
# F2
F2: <function indent at 0x7f48217ef1f0>
# F2
M1: <module 'navis.config' from '/home/jefferis/dev/Python/navis_error/navis/navis/config.py'>
D2: <dict object at 0x7f47ef5ae3c0>
D2: <dict object at 0x7f480c88f600>
D2: <dict object at 0x7f480ce7fe00>
# D2
# D2
M2: <module 'logging' from '/home/jefferis/.pyenv/versions/3.9.5/lib/python3.9/logging/__init__.py'>
# M2
M2: <module 'pint' from '/home/jefferis/.pyenv/versions/3.9.5/lib/python3.9/site-packages/pint/__init__.py'>
# M2
M2: <module 'matplotlib' from '/home/jefferis/.pyenv/versions/3.9.5/lib/python3.9/site-packages/matplotlib/__init__.py'>
# M2
F2: <function getLogger at 0x7f4821b861f0>
# F2
F2: <function default_logging at 0x7f480c8940d0>
# F2
F2: <function remove_log_handlers at 0x7f480c114700>
# F2
T4: <class 'pint.registry.UnitRegistry'>
# T4
D2: <dict object at 0x7f480bbaae80>
D2: <dict object at 0x7f480c898240>
T4: <class 'pint.context.Context'>
# T4
D2: <dict object at 0x7f480bbaaa80>
D2: <dict object at 0x7f480b9a0980>
T4: <class 'pint.util.UnitsContainer'>
# T4
T4: <class 'pint.util.udict'>
# T4
T4: <class 'pint.context.Expression'>
# T4
D2: <dict object at 0x7f480ba3f040>
# D2
D2: <dict object at 0x7f480ba3dd00>
# D2
D2: <dict object at 0x7f480ba3f240>
# D2
D2: <dict object at 0x7f480ba3f4c0>
# D2
D2: <dict object at 0x7f480ba3f580>
# D2
D2: <dict object at 0x7f480ba3f800>
# D2
D2: <dict object at 0x7f480ba3f8c0>
# D2
D2: <dict object at 0x7f480ba3f980>
# D2
D2: <dict object at 0x7f480ba3fa40>
# D2
D2: <dict object at 0x7f480ba3fac0>
# D2
D2: <dict object at 0x7f480ba3fb40>
# D2
D2: <dict object at 0x7f480ba3fbc0>
# D2
D2: <dict object at 0x7f480ba3fc40>
# D2
D2: <dict object at 0x7f480ba1f700>
# D2
D2: <dict object at 0x7f480ba3fcc0>
# D2
D2: <dict object at 0x7f480ba3fd00>
# D2
D2: <dict object at 0x7f480ba3fd40>
# D2
D2: <dict object at 0x7f480ba3fd80>
# D2
D2: <dict object at 0x7f480ba3fdc0>
# D2
D2: <dict object at 0x7f480ba3fe00>
# D2
D2: <dict object at 0x7f480ba3fe80>
# D2
D2: <dict object at 0x7f480ba3ff00>
# D2
D2: <dict object at 0x7f480ba3ff80>
# D2
D2: <dict object at 0x7f480ba24700>
# D2
D2: <dict object at 0x7f480ba45040>
# D2
D2: <dict object at 0x7f480ba45080>
# D2
D2: <dict object at 0x7f480ba45140>
# D2
D2: <dict object at 0x7f480ba45240>
# D2
D2: <dict object at 0x7f480ba45380>
# D2
D2: <dict object at 0x7f480ba454c0>
# D2
D2: <dict object at 0x7f480ba455c0>
# D2
D2: <dict object at 0x7f480ba45700>
# D2
# D2
D2: <dict object at 0x7f480bbaa980>
# D2
T4: <class 'weakref.WeakValueDictionary'>
# T4
D2: <dict object at 0x7f480b9ab140>
F1: <function WeakValueDictionary.__init__.<locals>.remove at 0x7f480bab0790>
Co: <code object remove at 0x7f4821bf8500, file "/home/jefferis/.pyenv/versions/3.9.5/lib/python3.9/weakref.py", line 106>
# Co
D4: <dict object at 0x7f4821bf7b40>
# D4
R1: <weakref at 0x7f480ba10b80; to 'WeakValueDictionary' at 0x7f480b9e1af0>
F2: <function _create_weakref at 0x7f481ebdd700>
# F2
# R1
D2: <dict object at 0x7f481ec00600>
# D2
D2: <dict object at 0x7f47f8af2540>
D2: <dict object at 0x7f481ec00640>
# D2
# D2
# F1
D2: <dict object at 0x7f480b9ab640>
```

Happy to dig a bit deeper on my end if that's required but thought I'd get your input first.

Thanks!

Philipp

**Comment by mmckerns** [comment]
Here's minimal code to show a `KeyedRef` object is not serializable with at least the most recent version of `dill` in python 3.10.4:
```
Python 3.10.4 (main, May 10 2022, 03:02:03) [Clang 10.0.1 (clang-1001.0.46.4)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import dill
>>> import weakref
>>> w = weakref.KeyedRef(weakref, None, None)
>>> dill.dumps(w)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/mmckerns/lib/python3.10/site-packages/dill-0.3.6.dev0-py3.10.egg/dill/_dill.py", line 364, in dumps
    dump(obj, file, protocol, byref, fmode, recurse, **kwds)#, strictio)
  File "/Users/mmckerns/lib/python3.10/site-packages/dill-0.3.6.dev0-py3.10.egg/dill/_dill.py", line 336, in dump
    Pickler(file, protocol, **_kwds).dump(obj)
  File "/Users/mmckerns/lib/python3.10/site-packages/dill-0.3.6.dev0-py3.10.egg/dill/_dill.py", line 620, in dump
    StockPickler.dump(self, obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 487, in dump
    self.save(obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 578, in save
    rv = reduce(self.proto)
TypeError: cannot pickle 'KeyedRef' object
>>> dill.__version__
'0.3.6.dev0'
```
are you suggesting that a `KeyedRef` is serializable using `dill-0.3.4` and earlier?

**Comment by schlegelp** [comment]
Thanks for the quick response! No, your minimal example also fails for me with `dill-0.3.4`.

All I can say at this point is that my unit tests worked just fine with `0.3.4` but throw a `KeyedRef` error with `0.3.5` 🤷 

I'll try to find a more concrete example.

**Comment by anivegesana** [comment]
Most likely, #443, which fixed a bug in function pickling, allowed more functions to be pickled correctly, which exposed more global variables which cannot be pickled. It is possible that they are not used by your functions. Try pickling with recurse mode on:

```py
dill.dumps(func, recurse=True)
```
or with the setting set globally
```py
dill.settings['recurse'] = True
```

**Comment by schlegelp** [comment]
The issue was likely some circular import between submodules. I managed to get it to behave by moving some decorators into separate modules and deferring a few imports.

I also tried to come up with a minimal working example based on these findings but couldn't reproduce it. I'm happy to close this issue but leave it up to you in case you'd like to follow up or wait for better description of the error.
